hazelcast/src/main/java/com/hazelcast/map/impl/MapKeyLoaderUtil.java

```diff
@@ -110,13 +112,19 @@ public boolean hasNext() {
      * until at least one group has up to {@code maxBatch}
      * entries or until the {@code entries} have been exhausted.
      *
-     * @param entries  the entries to be grouped by key
-     * @param maxBatch the maximum size of a group
+     * @param entries                  the entries to be grouped by key
+     * @param maxBatch                 the maximum size of a group
+     * @param nodeWideLoadedKeyLimiter controls the loaded number of keys per node
      * @return the grouped entries by entry key
      */
-    private static Map<Integer, List<Data>> nextBatch(Iterator<Entry<Integer, Data>> entries, int maxBatch) {
+    private static Map<Integer, List<Data>> nextBatch(Iterator<Entry<Integer, Data>> entries,
+                                                      int maxBatch, Semaphore nodeWideLoadedKeyLimiter) {
         Map<Integer, List<Data>> batch = createHashMap(maxBatch);
         while (entries.hasNext()) {
+            if (!nodeWideLoadedKeyLimiter.tryAcquire()) {
```

Comment:
I wonder if we should use `.acquire()` here or drop in an `IdleStrategy`. As I understand if this method returns an empty batch, we call this method in a tight loop until the semaphore gives us the green light to proceed. Along the path, we create hashmaps etc. 

As a second thought, shouldn't the semaphore be checked for available permits in `MapKeyLoader#sendKeysInBatches` in its 
```
while (batches.hasNext()) {
                Map<Integer, List<Data>> batch = batches.next();
                futures.addAll(sendBatch(batch, replaceExistingValues, nodeWideLoadedKeyLimiter));
            }
```
loop? Then the whole concurrency control would be in the same class and avoiding unnecessary allocations would be easier. WDYT?

Reply:
Acquire is not an option, since we have to return from method. But adding an idle-strategy makes sense. I have added it.